### PR TITLE
Improve shuffling performance

### PR DESCRIPTION
**Pull request by tbenr**
Avoids multiple `createSha256` and `Bytes.wrap` by introducing a `Sha256` class designed to be reused.

Before:
```
# Run complete. Total time: 00:01:26

Benchmark                            (indexCount)  Mode  Cnt  Score   Error  Units
ShuffleBenchmark.shuffledIndexBench         16384    ss   15  1.088 ± 0.017   s/op
ShuffleBenchmark.shuffledIndexBench         32768    ss   15  2.283 ± 0.185   s/op
ShuffleBenchmark.shuffledListBench          16384    ss   15  0.012 ± 0.004   s/op
ShuffleBenchmark.shuffledListBench          32768    ss   15  0.026 ± 0.003   s/op
```

After:
```
# Run complete. Total time: 00:01:17

Benchmark                            (indexCount)  Mode  Cnt  Score    Error  Units
ShuffleBenchmark.shuffledIndexBench         16384    ss   15  0.992 ±  0.020   s/op
ShuffleBenchmark.shuffledIndexBench         32768    ss   15  1.958 ±  0.021   s/op
ShuffleBenchmark.shuffledListBench          16384    ss   15  0.008 ±  0.001   s/op
ShuffleBenchmark.shuffledListBench          32768    ss   15  0.014 ±  0.001   s/op
```

Memory allocations comparison on shuffle benchmark:

image

image

related to: #6016

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/MiscHelpers.java [hunk 1]
```diff
@@ -39,6 +41,8 @@ public class MiscHelpers {
   // Math.pow(2, 8) - 1;
   public static final UInt64 MAX_RANDOM_BYTE = UInt64.valueOf(255);
 
+  static final byte[] EMPTY_HASH = Bytes.EMPTY.toArrayUnsafe();
```
Comment (rolfyone) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/MiscHelpers.java]:
```suggestion
  private static final byte[] EMPTY_HASH = Bytes.EMPTY.toArrayUnsafe();
```

Comment (tbenr) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/MiscHelpers.java]: yeah, that was my intention :)

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/MiscHelpers.java [hunk 2]
```diff
@@ -186,17 +190,25 @@ public void shuffleList(int[] input, Bytes32 seed) {
           flip = pivot - i;
           bitIndex = i & 0xff;
           if (bitIndex == 0 || i == mirror1) {
-            hashBytes = Hash.sha256(seed, roundAsByte, uintToBytes(i / 256, 4));
+            seed.update(digest);
+            roundAsByte.update(digest);
+            uintToBytes(i / 256, 4).update(digest);
+
+            hashBytes = digest.digest();
```
Comment (rolfyone) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/MiscHelpers.java]: these 2 blocks are basically still the sha256.. not against having our own implementation, but can we wrap it in a fn so it's still relatively easy to read?

**Comment by ajsutton** [comment]
Worth a changelog entry as well. Otherwise I agree with Paul about extracting a method to do the hash. Looks good otherwise.

**Comment by tbenr** [comment]
ouch, `computeProposerIndex` doesn't have a unit test

**Comment by tbenr** [comment]
I see I can apply the same pattern to:

`getNextSyncCommitteeIndices` and `isValidMerkleBranch`.

And maybe do some cleanup to ssz to use `Sha256` class

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/MiscHelpers.java [hunk 3]
```diff
@@ -79,15 +82,16 @@
...
-    Bytes32 hash = null;
+    byte[] hash = null;
     while (true) {
       int candidateIndex = indices.getInt(computeShuffledIndex(i % total, total, seed));
       if (i % 32 == 0) {
-        hash = Hash.sha256(seed, uint64ToBytes(Math.floorDiv(i, 32L)));
+        hash = sha256.digest(seed, uint64ToBytes(Math.floorDiv(i, 32L)));
       }
-      int randomByte = UnsignedBytes.toInt(hash.get(i % 32));
+      int randomByte = UnsignedBytes.toInt(hash[i % 32]);
```
[the remark this answers is not in the document]
Comment (tbenr) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/MiscHelpers.java:94]: this is not true

**Review by StefanBratanov — APPROVED** [review]
LGTM

**Pull request merged**